### PR TITLE
Ben add webhooks

### PR DESCRIPTION
**Pull request by benlower**

### CODE REVIEW

**Review by mdepinet — COMMENTED** [review]
Will review after other changes land and you get to merge main.

File: docs/src/content/docs/api/webhooks.mdx [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+  title="API Key Required"
+  text="The Ultravox API requires an API key. For more details see "
+  link={{ url: "../auth", text: "Authentication." }}
+/>
+<br />
+
+Webhooks allow you to receive real-time notifications about events in your Ultravox account. You can configure webhooks to send HTTP POST requests to a specified URL when certain events occur.
+
+## Webhooks in Ultravox
+When creating a webhook, you must provide a URL that will receive the webhook notification. Your service must return an HTTP status code of 200 ("OK"). If your service fails to respond with a 200, we will retry sending again.
```
Comment (mdepinet) [inline docs/src/content/docs/api/webhooks.mdx:20]: Any 200-class response code is fine (or 300-class for that matter).  Just not a 400- or 500-class.  (For example, return 204 No Content probably makes the most sense in general and would be totally fine.)

File: docs/src/content/docs/api/webhooks.mdx [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+### Securing Webhooks
+You can optionally choose to secure your webhooks with a key. When creating a webhook, a secret key is automatically generated for you or you can choose to provide your own secret. You can update or patch your webhooks to change secrets in the event of a leak or as part of regular key rotation.
+
+Each time your server receives an incoming webhook from Ultravox here's how you can ensure the webhook was sent by Ultravox and hasn't been tampered with:
+
+1. **Timestamp Verification**
+* Each incoming webhook request includes a `X-Ultravox-Webhook-Timestamp` header with the time the webhook was sent.
+* Verify that this timestamp is recent (e.g. within the last minute) to prevent replay attacks.
+
+2. **Signature Verification**
```
Comment (mdepinet) [inline docs/src/content/docs/api/webhooks.mdx:31]: I think it'd be worthwhile to include the Python code sample I sent you to be as specific as possible.  When I read through [ngrok's study on common webhook implementations](https://ngrok.com/blog-post/get-webhooks-secure-it-depends-a-field-guide-to-webhook-security#:~:text=Use%20asymmetric%20keys%20for%20non,fetch%20new%20keys%20as%20needed.) one of the things they found most frustrating was docs with only English or pseudocode. Anything other than actual code leaves enough ambiguity that a bit-by-bit comparison of a hash signature is likely to fail in some way that by its nature yields no information about why.

File: docs/src/content/docs/api/webhooks.mdx [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+* This allows for key rotation without downtime.
+* Your code should check if any of the provided signatures match your computed signature.
+
+By implementing these checks, you ensure that only authentic, recent, and unmodified webhooks from Ultravox are processed by your system.
+Remember to store your webhook secret securely and never expose it in client-side code or public repositories.
+
+### Retries
+When a destination URL fails to respond to an incoming webhook with a 200, we will retry using an exponential backoff strategy as follows:
+
+* First retry will occur 30 seconds later.
```
Comment (mdepinet) [inline docs/src/content/docs/api/webhooks.mdx:50]:
```suggestion
* First retry will occur approximately 30 seconds later.
```

File: docs/src/content/docs/api/webhooks.mdx [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+3. **Multiple Signatures**
+* `The X-Ultravox-Webhook-Signature` header may contain multiple signatures separated by commas.
+* This allows for key rotation without downtime.
+* Your code should check if any of the provided signatures match your computed signature.
+
+By implementing these checks, you ensure that only authentic, recent, and unmodified webhooks from Ultravox are processed by your system.
+Remember to store your webhook secret securely and never expose it in client-side code or public repositories.
+
+### Retries
+When a destination URL fails to respond to an incoming webhook with a 200, we will retry using an exponential backoff strategy as follows:
```
Comment (mdepinet) [inline docs/src/content/docs/api/webhooks.mdx:48]:
```suggestion
When a destination URL fails to respond to an incoming webhook with a 200, we will retry using a random exponential backoff strategy as follows:
```

File: docs/src/content/docs/api/webhooks.mdx (hunk 2 above)
Comment (mdepinet) [reply docs/src/content/docs/api/webhooks.mdx:31]: (Just as a simple example, this text doesn't mention that the date follows ISO formatting or that the signature is hex encoded)

**Review by mdepinet — APPROVED** [review]

File: docs/src/content/docs/api/webhooks.mdx (hunk 1 above)
Comment (benlower) [reply docs/src/content/docs/api/webhooks.mdx:20]: cool. adding this.

File: docs/src/content/docs/api/webhooks.mdx (hunk 2 above)
Comment (benlower) [reply docs/src/content/docs/api/webhooks.mdx:31]: yeah good idea. adding.

File: docs/src/content/codesnippets/webhooks-security.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import datetime
+import hmac
+
+request_timestamp = request.headers["X-Ultravox-Webhook-Timestamp"]
+if datetime.datetime.now() - datetime.dateimte.fromisoformat(request_timestamp) > datetime.timedelta(minutes=1):
```
Comment (mdepinet) [inline docs/src/content/codesnippets/webhooks-security.txt:5]: The last datetime here is misspelled as dateimte (which I assume was my fault typing this out in Slack :sweat_smile: )

**Pull request merged**